### PR TITLE
warm start updates and tied contact patch test

### PR DESCRIPTION
**Pull request by ebchin**
This PR adds penalty and Lagrange multiplier contact to the warm start.  Additionally, it adds a tied contact patch test that fails if warm start does not account for contact constraints.

Some other small changes:
- Variable initial steps in ironing example
- `ContactData::jacobianFunction()` was not using the displacement vector, so that argument has been removed from the method
- Support for a displacement boundary condition on one component specified by position and time
- Changed `ode_time_point_` to `time_` in `SolidMechanicsContact` to match what's in `SolidMechanics`

### CODE REVIEW

**Comment by ebchin** [comment]
This one should be ready to go!

File: src/serac/physics/solid_mechanics_contact.hpp [hunk 1]
```diff
@@ -249,10 +246,135 @@ class SolidMechanicsContact<order, dim, Parameters<parameter_space...>,
...
+      // use the most recently evaluated Jacobian
+      auto [_, drdu] = (*residual_)(time_, shape_displacement_, differentiate_wrt(displacement_), acceleration_,
+                                    *parameters_[parameter_indices].previous_state...);
+      J_.reset();
+      J_ = assemble(drdu);
+
+      contact_.update(cycle_, time_, dt);
+      if (contact_.haveLagrangeMultipliers()) {
+        J_offsets_    = mfem::Array<int>({0, displacement_.Size(), displacement_.Size() + contact_.numPressureDofs()});
+        J_constraint_ = contact_.jacobianFunction(J_.release());
```
Comment (tupek2) [inline src/serac/physics/solid_mechanics_contact.hpp:307]: this call does not take displacement.  Are we sure that the contact_ has the updated value for displacement internally at this point?

**Review by tupek2 — APPROVED** [review]
Looks fantastic.   This improves robustness for some problems?

File: src/serac/physics/solid_mechanics_contact.hpp (hunk 1 above)
Comment (ebchin) [reply src/serac/physics/solid_mechanics_contact.hpp:307]: Yes, the updated displacements are provided when `contact_.update()` is called on L309.  I added some preconditions to the doxygen comments in `contact_data.hpp` so it's clear you have to call `update()` with the current config if you want the Jacobian contributions to be accurate.

**Comment by ebchin** [comment]
> Looks fantastic. This improves robustness for some problems?

Yeah, this helps.  It cuts down on the number of Newton iterations needed for all of the contact examples as compared to turning warm start off.

File: src/serac/physics/solid_mechanics.hpp [hunk 2]
```diff
@@ -459,6 +459,27 @@ class SolidMechanics<order, dim, Parameters<parameter_space...>, std::integer_se
...
+   * @brief Set the displacement essential boundary conditions on a single component
+   *
+   * @param[in] disp_bdr The set of boundary attributes to set the displacement on
+   * @param[in] disp The vector function containing the set displacement values
+   * @param[in] component The component to set the displacment on
+   *
+   * For the displacement function, the argument is the input position, the second argument is the time, and the output
+   * is the value of the component of the displacement.
+   *
+   * @note This method must be called prior to completeSetup()
```
Comment (white238) [inline src/serac/physics/solid_mechanics.hpp:472]: Can/should this be checked at runtime?

File: src/serac/physics/solid_mechanics_contact.hpp [hunk 3]
```diff
@@ -246,10 +241,135 @@ class SolidMechanicsContact<order, dim, Parameters<parameter_space...>,
...
+        J_operator_ = J_.get();
+      }
+
+      // Update the linearized Jacobian matrix
+      mfem::Vector augmented_residual(displacement_.space().TrueVSize() + contact_.numPressureDofs());
+      augmented_residual     = 0.0;
+      const mfem::Vector res = (*residual_)(time_ + dt, shape_displacement_, displacement_, acceleration_,
+                                            *parameters_[parameter_indices].state...);
+
+      // TODO this copy is required as the sundials solvers do not allow move assignments because of their memory
```
Comment (white238) [inline src/serac/physics/solid_mechanics_contact.hpp:342]: Is this a TODO or a NOTE?

File: src/serac/physics/tests/contact_patch_tied.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  StateManager::initialize(datastore, name + "_data");
+
+  // Construct the appropriate dimension mesh and give it to the data store
+  std::string filename = SERAC_REPO_DIR "/data/meshes/twohex_for_contact.mesh";
+
+  auto mesh = mesh::refineAndDistribute(buildMeshFromFile(filename), 3, 0);
+  StateManager::setMesh(std::move(mesh), "patch_mesh");
+
+// #ifdef SERAC_USE_PETSC
+//   LinearSolverOptions linear_options{
```
Comment (white238) [inline src/serac/physics/tests/contact_patch_tied.cpp:49]: Should this be removed or a note added why its commented out?

File: src/serac/physics/tests/CMakeLists.txt [hunk 5]
```diff
@@ -41,7 +41,11 @@ set(physics_parallel_test_sources
     solid_reaction_adjoint.cpp
     thermal_nonlinear_solve.cpp
     )
-blt_list_append(TO physics_parallel_test_sources ELEMENTS contact_patch.cpp contact_beam.cpp IF TRIBOL_FOUND)
+blt_list_append(TO physics_parallel_test_sources ELEMENTS 
+    contact_patch.cpp
+    contact_patch_tied.cpp
+    contact_beam.cpp
+    IF TRIBOL_FOUND)
```
Comment (white238) [inline src/serac/physics/tests/CMakeLists.txt]:
```suggestion
blt_list_append(TO physics_parallel_test_sources
                ELEMENTS 
                   contact_patch.cpp
                   contact_patch_tied.cpp
                   contact_beam.cpp
                IF TRIBOL_FOUND)
```

**Review by white238 — APPROVED** [review]
minor suggestions but thanks for the fixes!

File: src/serac/physics/solid_mechanics.hpp (hunk 2 above)
Comment (ebchin) [reply src/serac/physics/solid_mechanics.hpp:472]: Yeah, that's a good idea.  @btalamini is doing a more complete overhaul of this now.  Could this check be added to your Dirichlet BC overhaul, Brandon?

File: src/serac/physics/solid_mechanics_contact.hpp (hunk 3 above)
Comment (ebchin) [reply src/serac/physics/solid_mechanics_contact.hpp:342]: Probably a TODO once mfem/mfem#3531 is addressed, but it doesn't look like that will happen in the near term.  I'll change it to a NOTE

File: src/serac/physics/tests/contact_patch_tied.cpp (hunk 4 above)
Comment (ebchin) [reply src/serac/physics/tests/contact_patch_tied.cpp:49]: Note's been added.  We should add PETSc back in, but the convergence was slow with these settings.

File: src/serac/physics/solid_mechanics.hpp (hunk 2 above)
Comment (tupek2) [reply src/serac/physics/solid_mechanics.hpp:472]: We have these comments all over the place.  At some point we can go through and have a check that people are not calling this after completeSetup(), but I think its out of the scope for this.

**Pull request merged**